extensions/amp-analytics/0.1/transport.js

```diff
@@ -118,13 +144,272 @@ export class Transport {
+      return;
+    }
+    ampDoc.body.removeChild(Transport.crossDomainIframes_[frameUrl].frame);
+    delete Transport.crossDomainIframes_[frameUrl];
+  }
+
+  /**
+   * Record that one more creative is using this cross-domain iframe
+   * @param {!string} frameUrl The URL of the frame
+   * @return {number}
+   * @private
+   */
+  static incrementIframeUsageCount_(frameUrl) {
+    return ++(Transport.crossDomainIframes_[frameUrl].usageCount);
+  }
+
+  /**
+   * Record that one fewer creative is using this cross-domain iframe
+   * @param {!string} frameUrl The URL of the frame
+   * @return {number}
+   * @private
+   */
+  static decrementIframeUsageCount_(frameUrl) {
+    return --(Transport.crossDomainIframes_[frameUrl].usageCount);
+  }
+
+  /**
+   * Returns whether a url of a cross-domain frame is already known
+   * @param {!string} frameUrl
+   * @return {!boolean}
+   * @private
+   */
+  static hasCrossDomainIframe_(frameUrl) {
+    return hasOwn(Transport.crossDomainIframes_, frameUrl);
+  }
+
+  /**
+   * Create a cross-domain iframe for third-party vendor anaytlics
+   * @param {!HTMLDocument} ampDoc  The document node of the parent page
+   * @param {!string} frameUrl  The URL of the cross-domain iframe
+   * @param {string=} opt_extraData The data to send to the frame
+   * @return {!Element}
+   */
+  createCrossDomainIframe_(ampDoc, frameUrl, opt_extraData) {
+    const sentinel = Transport.createUniqueId_();
+    const scriptSrc = getMode().localDev
+      ? '/dist.3p/current/ampanalytics-lib.js'
+      : `${urls.thirdParty}/$internalRuntimeVersion$/ampanalytics-v0.js`;
+    const frame = createElementWithAttributes(ampDoc, 'iframe', {
+      sandbox: 'allow-scripts',
+      name: JSON.stringify({
+        scriptSrc,
+        sentinel,
+      }),
+    });
+    const iframeMessagingClient = new IframeMessagingClient(window);
+    iframeMessagingClient.setSentinel(sentinel);
+    iframeMessagingClient.setHostWindow(frame);
+    setStyles(frame,
+        {width: 0, height: 0, display: 'none',
+         position: 'absolute', top: 0, left: 0});
+    Transport.crossDomainIframes_[frameUrl] = {
+      frame,
+      sentinel,
+      isReady: false,
+      messageQueue: [],
+      usageCount: 1,
+      send: messages => {
+        const envelope = {};
+        envelope[MessageTypes.ampAnalytics3pMessages] = messages;
+        iframeMessagingClient.sendMessage(MessageTypes.ampAnalytics3pMessages,
+          envelope);
+      },
+      iframeMessagingClient,
+      sendTimer: null,
+    };
+    this.informCrossDomainIframeOfNewCreative_(frameUrl, opt_extraData);
+    frame.src = frameUrl;
+    return frame;
+  }
+
+  /**
+   * Create a unique value to differentiate messages from
+   * this particular creative to the cross-domain iframe
+   * @returns {string}
+   * @private
+   */
+  static createUniqueId_() {
+    let newId;
+    do {
+      newId = String(Math.random()).substr(2);
+    } while (Transport.usedIds_[newId]);
```

Comment:
I'm hoping we move to incrementing number but assuming we don't, I would argue the probability of an inadvertent, neverending loop is higher than 20x random not creating unique ID and as mentioned, you could append the type (e.g. moat) to ensure its truly unique.